nixos/containers: mtu setting for veths

benaryorg opened the pull request:
Allowing to specify the MTU for veths makes it possible to use jumbo frames, as well as to forego MSS clamping in setups where either is needed. To be noted here is that the MTU for veths is in fact relevant, unlike the loopback interface (which is weird about MTUs), or e.g. the link speed which seems to be purely CPU bound for veths. Since MTUs are a relatively common thing (VPNs being ubiquitous these days) which can cause connectivity issues when MSS clamping isn't performed (something that works *most* of the time, but which many people may not be aware of) this options seems warranted.

Since the conditions for setting up the interface in most places are that an address is provided either on the host side or the guest side, this change enables using that mechanic but without an address. This can be useful in environment where the interface is configured via a network manager on both sides of the pair. Basically; setting the MTU alone is now sufficient to have the NixOS scripts handle the interface, but not configure addresses or routes on it, in case a user wants to manage those settings themselves. For everyone using the address built-in mechanics it merely allows to specify the MTU.

The envvar for private users in the nspawn call is moved up so that the network options are grouped together.


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

benaryorg committed:
nixos/containers: mtu setting for veths

Allowing to specify the MTU for veths makes it possible to use jumbo frames, as well as to forego MSS clamping in setups where either is needed.
To be noted here is that the MTU for veths is in fact relevant, unlike the loopback interface (which is weird about MTUs), or e.g. the link speed which seems to be purely CPU bound for veths.
Since MTUs are a relatively common thing (VPNs being ubiquitous these days) which can cause connectivity issues when MSS clamping isn't performed (something that works *most* of the time, but which many people may not be aware of) this options seems warranted.

Since the conditions for setting up the interface in most places are that an address is provided either on the host side or the guest side, this change enables using that mechanic but without an address.
This can be useful in environment where the interface is configured via a network manager on both sides of the pair.
Basically; setting the MTU alone is now sufficient to have the NixOS scripts handle the interface, but not configure addresses or routes on it, in case a user wants to manage those settings themselves.
For everyone using the address built-in mechanics it merely allows to specify the MTU.

The envvar for private users in the nspawn call is moved up so that the network options are grouped together.

Signed-off-by: benaryorg <binary@benary.org>

benaryorg commented:
Local tests are still churning, and I'll see if I can add a quick check/test for the MTU too.

benaryorg committed:
nixos/containers: tests for MTU functionality

The tests are pretty straightforward; create containers for IPv6 and IPv4 each with MTUs at 1000, 1500, 9000, and no MTU, and then ping them with appropriate link layer sizes.

Signed-off-by: benaryorg <binary@benary.org>

benaryorg commented:
Not sure who to ping/request reviews from here.
@kampfschlaefer seems to be maintaining a lot of container tests so here's a ping.

I'll let this PR sit for a bit and ping some more people or request reviews at a later time if nothing happens.
This isn't urgent after all ^^

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/5536

benaryorg commented:
quick rebase

benaryorg commented:
Ah, my bad, resolved the merge conflict lexicographically wrong. Should be fixed now.

benaryorg commented:
Refactored the entire PR to the new quoting style (i.e. more bashisms), and also removed a deprecation warning.
The test is passing just fine (and properly caught an error I introduced in the refactoring).

ping @ElvishJerricco @Ma27 (going by the people mentioned in other PRs on the same files)

benaryorg commented:
Quick rebase.

By the way this has been in production on my systems for almost a year, I think it's safe to merge ^^
